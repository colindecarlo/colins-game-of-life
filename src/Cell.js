class Cell {
  constructor() {
    this.animated = false;
    this.neighbours = [];
  }

  static dead() {
    return new Cell;
  }

  tick() {
    this.animate();
  }

  isDead() {
    return this.animated == false;
  }

  animate() {
    this.animated = true;
  }

  addNeighbours(newNeighbours) {
    this.neighbours = [...this.neighbours, ...newNeighbours];
  }
}

export default Cell