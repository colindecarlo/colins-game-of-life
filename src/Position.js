export default class Position {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }

  static fromString(position) {
    const coords = /\((-?\d+), (-?\d+)\)/.exec(position);
    return new Position(parseInt(coords[1], 10), parseInt(coords[2], 10));
  }

  topLeft() {
    return new Position(this.x - 1, this.y - 1);
  }

  top() {
    return new Position(this.x, this.y - 1);
  }

  topRight() {
    return new Position(this.x + 1, this.y - 1);
  }

  left() {
    return new Position(this.x - 1, this.y);
  }

  right() {
    return new Position(this.x + 1, this.y);
  }

  bottomLeft() {
    return new Position(this.x - 1, this.y + 1);
  }

  bottom() {
    return new Position(this.x, this.y + 1);
  }

  bottomRight() {
    return new Position(this.x + 1, this.y + 1);
  }

  neighbouringPositions() {
    return [
      this.topLeft(),
      this.top(),
      this.topRight(),

      this.left(),
      this.right(),

      this.bottomLeft(),
      this.bottom(),
      this.bottomRight(),
    ];
  }

  toString() {
    return `(${this.x}, ${this.y})`;
  }
}
