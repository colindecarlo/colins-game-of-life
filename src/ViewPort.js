export default class ViewPort {
  constructor(topLeft, bottomRight) {
    this.topLeft = topLeft;
    this.bottomRight = bottomRight;
  }
}
