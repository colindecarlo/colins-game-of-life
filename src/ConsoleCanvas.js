import Position from './Position';

export default class ConsoleCanvas {
  print(world, viewPort) {
    const canvas = [];
    for (let { y } = viewPort.topLeft; y <= viewPort.bottomRight.y; y++) {
      const row = [];
      for (let { x } = viewPort.topLeft; x <= viewPort.bottomRight.x; x++) {
        const representation = world.cellAt(new Position(x, y)).isAnimate() ? 'o' : ' ';
        row.push(representation);
      }
      canvas.push(row);
    }

    return canvas;
  }
}
