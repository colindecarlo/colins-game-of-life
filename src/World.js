import CollisionError from './CollisionError';
import Position from './Position';
import Cell from './Cell';
import ViewPort from './ViewPort';

class World {
  constructor() {
    this.grid = new Map();
  }

  get cells() {
    return Array.from(this.grid.values());
  }

  get boundaries() {
    const positions = Array.from(this.grid.keys(), position => Position.fromString(position));
    const toSmallest = (smallest, value) => Math.min(smallest, value);
    const toBiggest = (biggest, value) => Math.max(biggest, value);

    const topMost = positions.map(position => position.y).reduce(toSmallest);
    const leftMost = positions.map(position => position.x).reduce(toSmallest);
    const bottomMost = positions.map(position => position.y).reduce(toBiggest);
    const rightMost = positions.map(position => position.x).reduce(toBiggest);

    return new ViewPort(new Position(leftMost - 1, topMost - 1), new Position(rightMost + 1, bottomMost + 1));
  }

  seed(cells) {
    cells.forEach((cell) => {
      this.seedCell(cell, this.emptyPositionInRange());
    });
  }

  tick() {
    const nextGeneration = new Map();
    const viewPort = this.boundaries;
    for (let { y } = viewPort.topLeft; y <= viewPort.bottomRight.y; y++) {
      for (let { x } = viewPort.topLeft; x <= viewPort.bottomRight.x; x++) {
        const position = new Position(x, y);
        const cell = this.cellAt(position);

        const nextGenCell = new Cell();
        if (cell.isAnimate()) {
          nextGenCell.animate();
        }

        nextGenCell.addNeighbours(this.neighboursOf(position));
        nextGenCell.tick();
        nextGenCell.removeNeighbours();

        if (nextGenCell.isAnimate()) {
          nextGeneration.set(position.toString(), nextGenCell);
        }
      }
    }

    this.grid = nextGeneration;
  }

  emptyPositionInRange() {
    let position;
    do {
      position = this.positionInRange();
    } while (this.grid.has(position.toString()));
    return position;
  }

  seedCell(cell, position) {
    const key = position.toString();

    if (this.grid.has(key)) {
      throw new CollisionError(this.grid.get(key), cell);
    }

    this.grid.set(key, cell);
  }

  cellAt(position) {
    const key = position.toString();
    return this.grid.get(key) || Cell.dead();
  }

  positionInRange() {
    const x = Math.floor(Math.random() * 15);
    const y = Math.floor(Math.random() * 15);

    return new Position(x, y);
  }

  neighboursOf(position) {
    return [
      this.grid.get(position.topLeft().toString()) || Cell.dead(),
      this.grid.get(position.top().toString()) || Cell.dead(),
      this.grid.get(position.topRight().toString()) || Cell.dead(),

      this.grid.get(position.left().toString()) || Cell.dead(),
      this.grid.get(position.right().toString()) || Cell.dead(),

      this.grid.get(position.bottomLeft().toString()) || Cell.dead(),
      this.grid.get(position.bottom().toString()) || Cell.dead(),
      this.grid.get(position.bottomRight().toString()) || Cell.dead(),
    ];
  }
}

export default World;
