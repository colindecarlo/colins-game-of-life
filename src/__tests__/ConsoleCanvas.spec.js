import Position from '../Position';
import Cell from '../Cell';
import World from '../World';
import ConsoleCanvas from '../ConsoleCanvas';
import ViewPort from '../ViewPort';

describe('Console Canvas', () => {
  it('can render the world', () => {
    const glider = [
      new Position(1, 0),
      new Position(2, 1),
      new Position(0, 2),
      new Position(1, 2),
      new Position(2, 2),
    ];

    const world = new World();
    glider.forEach(position => world.seedCell(Cell.alive(), position));

    const canvas = new ConsoleCanvas();

    const viewPort = new ViewPort(new Position(-4, -4), new Position(5, 5));
    const view = canvas.print(world, viewPort);

    expect(view.length).toBe(10);
    expect(view).toMatchSnapshot();
  });
});
