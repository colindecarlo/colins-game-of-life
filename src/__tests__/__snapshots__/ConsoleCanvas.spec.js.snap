// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Console Canvas can render the world 1`] = `
Array [
  Array [
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
  ],
  Array [
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
  ],
  Array [
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
  ],
  Array [
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
  ],
  Array [
    " ",
    " ",
    " ",
    " ",
    " ",
    "o",
    " ",
    " ",
    " ",
    " ",
  ],
  Array [
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    "o",
    " ",
    " ",
    " ",
  ],
  Array [
    " ",
    " ",
    " ",
    " ",
    "o",
    "o",
    "o",
    " ",
    " ",
    " ",
  ],
  Array [
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
  ],
  Array [
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
  ],
  Array [
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
    " ",
  ],
]
`;
